### PR TITLE
V3 layout

### LINKED ISSUE #30: Dashboard view layout
As a user I need a dashboard view where to manage the exhibitions, devices, layouts, view current live exhibitions and view analytics

### PR DESCRIPTION
**Pull request by lauriLukkarinen**
Resolves #30

### CODE REVIEW

File: src/components/dashboard/dashboard-devices-view.tsx [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  }
+
+  /**
+   * Component render method
+   */
+  public render() {
+    const { classes, history } = this.props;
+
+    const filterOptions = [
+      { name: "Kaikki", value: "all" }
```
Comment (anttileppa) [inline src/components/dashboard/dashboard-devices-view.tsx]: Localize

File: src/components/dashboard/dashboard-layout.tsx [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    return (
+      <div className={ classes.root }>
+        <header className={ classes.header }>
+          {
+            this.renderHeader()
+          }
+        </header>
+        <div className={ classes.navigation }>
+          <div className={ classes.userElement }>
+              <div className={ classes.userAvatar }>TR</div>
```
Comment (anttileppa) [inline src/components/dashboard/dashboard-layout.tsx]: First and last name from logged user?

File: src/components/dashboard/dashboard-layout.tsx [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+              <h3>Timo</h3>
```
Comment (anttileppa) [inline src/components/dashboard/dashboard-layout.tsx]: This value should be taken from access token

File: src/components/dashboard/dashboard-recent-view.tsx [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+  private sortExhibitions = (exhibitions: Exhibition[]) => {
+    const { recentSortBy } = this.state;
+    switch (recentSortBy) {
+      case ExhibitionRecentSortBy.MODIFIED_AT: {
+        return exhibitions.sort((a, b) => {
+          if (!a.modifiedAt || !b.modifiedAt) {
+            return 0;
+          }
+
+          return moment(b.modifiedAt).unix() - moment(a.modifiedAt).unix();
```
Comment (anttileppa) [inline src/components/dashboard/dashboard-recent-view.tsx]: Use moment's query API https://momentjs.com/docs/#/query/

File: src/components/dashboard/dashboard-recent-view.tsx [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+        });
+      }
+      case ExhibitionRecentSortBy.CREATED_AT: {
+        return exhibitions.sort((a, b) => {
+          if (!a.createdAt || !b.createdAt) {
+            return 0;
+          }
+
+          return moment(b.createdAt).unix() - moment(a.createdAt).unix();
```
Comment (anttileppa) [inline src/components/dashboard/dashboard-recent-view.tsx]: Use moment's query API https://momentjs.com/docs/#/query/

File: src/components/exhibition-v3/exhibition-view-v3.tsx [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            <TreeView>
+              <TreeItem nodeId={ exhibition.id } label={ exhibition.name }>
+                { exhibitionRooms &&
+                  exhibitionRooms.map(room => this.renderExhibitionRoomTreeItem(room))
+                }
+              </TreeItem>
+            </TreeView>
+          </ElementNavigationPane>
+          <EditorView>
+            <Typography>Olen näyttelyeditorin v3 sisältö { history.location.pathname } </Typography>
```
Comment (anttileppa) [inline src/components/exhibition-v3/exhibition-view-v3.tsx]: Remove this silly text

File: src/components/exhibition-v3/exhibition-view-v3.tsx [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+          </EditorView>
+          <ElementSettingsPane title="2. Huone 1" />
```
Comment (anttileppa) [inline src/components/exhibition-v3/exhibition-view-v3.tsx]: Remove this placeholder text

File: src/components/layout/layout-editor-view.tsx [hunk 8]
```diff
@@ -0,0 +1,603 @@
...
+      <div className={ classes.root }>
+        <div className={ classes.panel } style={{ width: this.state.toolbarOpen ? minWidth : minimizedWidth }}>
+          <div className={ classes.btnContainer }>
+            <IconButton size="small" edge="start" onClick={ this.onToggleClick }>
+              { this.state.toolbarOpen ? <CloseIcon /> : <OpenIcon /> }
+            </IconButton>
+          </div>
+          <div className={ classNames( classes.container, this.state.toolbarOpen ? "" : "closed" ) }>
+            <div className={ classes.header }>
+              <h3>Layout</h3>
```
Comment (anttileppa) [inline src/components/layout/layout-editor-view.tsx]: Localize this

File: src/components/layout/layout-editor-view.tsx [hunk 9]
```diff
@@ -0,0 +1,603 @@
...
+  /**
+   * Renders a toolbar
+   */
+  private renderToolbar = () => {
+    const { classes } = this.props;
+
+    return (
+      <div className={ classes.toolBar }>
+        <Select value={ this.state.view } onChange={ this.onViewChange }>
+          <MenuItem value={ "CODE" }> Code </MenuItem>
```
Comment (anttileppa) [inline src/components/layout/layout-editor-view.tsx]: Localize these

File: src/components/layout/layout-editor-view.tsx [hunk 10]
```diff
@@ -0,0 +1,577 @@
...
+        <DialogActions>
+          <Button onClick={ this.onDeleteDialogCancelButtonClick } color="primary">{ strings.exhibitionLayouts.editView.deleteConfirmCancel }</Button>
+          <Button onClick={ this.onDeleteDialogDeleteButtonClick } color="primary" autoFocus>{ strings.exhibitionLayouts.editView.deleteConfirmDelete }</Button>
+        </DialogActions>
+      </Dialog>
+    );
+  }
+
+  private updateEditorData = () => {
+    const { layout } = this.props;
```
Comment (MrWalrus3451) [inline src/components/layout/layout-editor-view.tsx]: Missing docs.

File: src/components/exhibition-v3/exhibition-view-v3.tsx [hunk 11]
```diff
@@ -0,0 +1,871 @@
...
+          className={ classes.editor }
+          value={ this.state.jsonCode }
+          options={ jsonEditorOptions }
+          onBeforeChange={ this.onBeforeJsonCodeChange } />
+      </div>
+    );
+  }
+
+  /**
+   * Renders element settings pane content
```
[the remark this answers is not in the document]
Comment (lauriLukkarinen) [reply src/components/exhibition-v3/exhibition-view-v3.tsx]: actually pane is correct :) it means one separated part of the application window the user can interact with

File: src/components/exhibition-v3/exhibition-view-v3.tsx [hunk 12]
```diff
@@ -0,0 +1,871 @@
...
+
+    return (
+      <Select fullWidth value={ selectedElement.layoutId } onChange={ this.onLayoutChange }>
+        { layoutSelectItems }
+      </Select>
+    );
+  }
+
+  /**
+   * Remders resources tree
```
[the remark this answers is not in the document]
Comment (lauriLukkarinen) [reply src/components/exhibition-v3/exhibition-view-v3.tsx]: Fixed

File: src/components/layout/layout-editor-view.tsx (hunk 10 above)
Comment (lauriLukkarinen) [reply src/components/layout/layout-editor-view.tsx]: Fixed

**Pull request merged**